### PR TITLE
test: add data-driven Avro decode integration tests

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Background:
- Our test for different encodings is badly organized https://github.com/risingwavelabs/risingwave/issues/16931. And we mainly use e2e tests for them.
- Recently we have quite some features & bug fixes for them. Testing is one major obstacle of these work. It's now full of workarounds and mysterious conventions. 
	- #17110
	- #16948
	- #17069 
	- #17052
	- #16941 
	- #16273 

This PR proposes to improve the tests with 2 ideas:

1. Do unit/integration test for encodings. 
   
   Logically, these features have a nice clear boundary, there's no reason to start a cluster and write verbose e2e test for them, except we don't have a clear boundary in code. But we are improving the code organization during #17002, and the test boundary can also be improved now.

2. Data-driven tests.
   
   Like planner test, execution integration test... We just write input data (instead of using code to generate data). The result can be **generated**, and displayed in nice human readable format.

Benefits:
- Faster iteration loop: faster compilation, less steps to run tests: Just `UPDATE_EXPECT=1 cargo test -p risingwave_connector_codec`, or use Rust Analyzer to `Run test + expect`
- Easier to write tests, and then higher coverage.

Limitations:
- The boundary still isn't clear enough, and we need to re-implement some logic in tests. For more, see my comments in code.
- Avro rust lib is pooooor, so the "data driven" part isn't good enough.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
* **#17434** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17434?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/connector/src/parser/avro/parser.rs [hunk 1]
```diff
@@ -261,290 +251,4 @@ mod test {
...
-    #[tokio::test]
-    async fn update_avro_payload() {
-        let conf = new_avro_conf_from_local("simple-schema.avsc")
-            .await
-            .unwrap();
-        let mut writer = Writer::new(conf.schema.original_schema.as_ref(), Vec::new());
-        let record = build_avro_data(conf.schema.original_schema.as_ref());
-        writer.append(record).unwrap();
-        let encoded = writer.into_inner().unwrap();
-        println!("path = {:?}", e2e_file_path("avro_simple_schema_bin.1"));
```
Comment (xxchan) [inline src/connector/src/parser/avro/parser.rs:536]: There're still some usage of this data file to be cleaned.
image

File: src/connector/src/parser/avro/parser.rs [hunk 2]
```diff
@@ -261,290 +251,4 @@ mod test {
...
-    #[tokio::test]
-    async fn test_new_avro_parser() {
-        let avro_parser_rs = new_avro_parser_from_local("simple-schema.avsc").await;
-        let avro_parser = avro_parser_rs.unwrap();
-        println!("avro_parser = {:?}", avro_parser);
-    }
-
-    #[tokio::test]
-    async fn test_avro_union_type() {
-        let parser = new_avro_parser_from_local("union-schema.avsc")
```
Comment (fuyufjh) [inline src/connector/src/parser/avro/parser.rs:456]: Why this case is not in the new tests?

**Review by fuyufjh — COMMENTED** [review]
LGTM, but I didn't really get the motivation. Perhaps @xiangjinwu knows better 😄

File: src/connector/src/parser/avro/parser.rs (hunk 2 above)
Comment (xxchan) [reply src/connector/src/parser/avro/parser.rs:456]: You are right. It seems to be missed.

File: src/connector/codec/tests/integration_tests/utils.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+pub struct DataTypeTestDisplay<'a>(pub &'a DataType);
+
+impl<'a> std::fmt::Debug for DataTypeTestDisplay<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self.0 {
+            DataType::Struct(s) => {
+                // since this is for tests, we don't care about perf here :)
+                if !f.alternate() || s.len() == 1 {
+                    let (name, ty) = s.iter().next().unwrap();
+                    return write!(f, "Struct {{ {}: {:?} }}", name, &DataTypeTestDisplay(ty));
```
Comment (xiangjinwu) [inline src/connector/codec/tests/integration_tests/utils.rs:33]: Why do we only display the first field of a struct?

**Review by xiangjinwu — APPROVED** [review]
LGTM!

File: src/connector/codec/tests/integration_tests/utils.rs (hunk 3 above)
Comment (xxchan) [reply src/connector/codec/tests/integration_tests/utils.rs:33]: 😄

**Pull request merged**